### PR TITLE
[PT-D][DDP] Add unit tests for Forward and Backward Hook

### LINKED ISSUE #35191: Forward/backward hooks support in DistributedDataParallel
## 🚀 Feature
PyTorch now recommends to use DistributedDataParallel over DataParallel for all sorts of multi-GPU trainings (#35063). However, it has one limitation comparing to old DataParallel module - currently it cannot handle forward/backward hooks in a user convenient way.
Proposed workaround
https://github.com/pytorch/pytorch/blob/95ad94c75b09ad2438141e4eb52e83e737966e60/torch/nn/parallel/distributed.py#L146-L149
requires users to edit each model's forward propagation code for using hooks with model wrapped into DDP.
As I understand, it wasn't initially designed having this limitation in mind and was discovered during fixing another issue #5061. So, I am wondering, maybe there are some possibilities to implement some sort of hook synchronization mechanism across distributed model replicas?

## Motivation

Also with current workaround possibilities to use hooks dynamically is lost for DistributedDataParallel module. For example, in my current code with DataParallel I am able to place and remove hooks dynamically: during validation phase of training process I am placing hooks to extract additional bottleneck features to calculate some complementary evaluation metrics which are not calculated during training phase.
In general, current hooking mechanism looks not fully compatible with DDP.

## Pitch

Hooking mechanism for DistributedDataParallel module working from the user perspective as in DataParallel module.

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74063

Address the issue https://github.com/pytorch/pytorch/issues/66229 as part of BE effort.

Basically:
1. We remove the stale comment which confuses users.
2. Add more unit tests to test the forward/backward hook working for DDP.

Differential Revision: [D34800830](https://our.internmc.facebook.com/intern/diff/D34800830/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34800830/)!

### CODE REVIEW

File: test/distributed/nn/test_ddp_hook.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        except OSError:
+            pass
+
+    @property
+    def world_size(self):
+        return 2
+
+    @requires_nccl()
+    @skip_if_lt_x_gpu(2)
+    def test_ddp_hook(self):
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/nn/test_ddp_hook.py:69]: sure.

[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/nn/test_ddp_hook.py:69]: Do you feel we also need this test for gloo or we think nccl is enough?

**Comment by fduwjj** [comment]
Move the test to the suggested test file.

**Review by rohan-varma — APPROVED** [review]
awesome, thanks!

**Pull request closed without merging**